README.md

![stability-wip](https://img.shields.io/badge/stability-work_in_progress-lightgrey.svg)
[![Travis-CI Build
Status](https://travis-ci.org/hrbrmstr/porc.svg?branch=master)](https://travis-ci.org/hrbrmstr/porc)
[![Coverage
Status](https://img.shields.io/codecov/c/github/hrbrmstr/porc/master.svg)](https://codecov.io/github/hrbrmstr/porc?branch=master)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/hrbrmstr/porc?branch=master&svg=true)](https://ci.appveyor.com/project/hrbrmstr/porc)

# porc

Tools to Work with ‘Snort’ Rules, Logs and Data

## Description

‘Snort’ is an open source intrusion prevention system capable of
real-time traffic analysis and packet logging. Tools are provided to
work with ‘Snort’ rulesets, logs and other data associated with the
platform. More information on ‘Snort’ can be found at
<https://www.snort.org/>.

## What’s Inside The Tin

The following functions are implemented:

  - `download_community_rules`: Download Snort community rules
  - `download_subscription_rules`: Download Snort subscription rules
  - `expand_port_ranges`: Expand a Snort port description
  - `read_rules`: Parse in a file of snort rules into a data frame
  - `read_extended`: Read a Snort “extended v2” log into a data frame
  - `rule_vars`: Extract all the ‘$’-named variables from Snort rules
  - `as_rule_df`: Helper to class a Snort rules data frame properly

## Installation

``` r
devtools::install_github("hrbrmstr/porc")
```

## Usage

``` r
library(porc)
library(tidyverse)

# current verison
packageVersion("porc")
```

    ## [1] '0.1.1'

## Basic rule reading

``` r
rules <- read_rules(system.file("extdata", "emerging-telnet.rules", package="porc"))

glimpse(rules)
```

    ## Observations: 13
    ## Variables: 10
    ## $ action    <chr> "alert", "alert", "alert", "alert", "alert", "alert", "alert", "alert", "alert", "alert", "alert"...
    ## $ protocol  <chr> "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp"
    ## $ src_addr  <chr> "$HOME_NET", "$HOME_NET", "$HOME_NET", "$TELNET_SERVERS", "$TELNET_SERVERS", "$TELNET_SERVERS", "...
    ## $ src_ports <chr> "23", "23", "23", "23", "23", "23", "23", "any", "any", "any", "any", "any", "any"
    ## $ direction <chr> "unidirectional", "unidirectional", "unidirectional", "unidirectional", "unidirectional", "unidir...
    ## $ dst_addr  <chr> "$EXTERNAL_NET", "$EXTERNAL_NET", "$EXTERNAL_NET", "$EXTERNAL_NET", "$EXTERNAL_NET", "$EXTERNAL_N...
    ## $ dst_ports <chr> "any", "any", "any", "any", "any", "any", "any", "[23,2323,3323,4323]", "[23,2323,3323,4323]", "[...
    ## $ commented <lgl> FALSE, TRUE, FALSE, TRUE, FALSE, FALSE, FALSE, FALSE, TRUE, TRUE, FALSE, FALSE, FALSE
    ## $ options   <list> [<# A tibble: 9 x 2,      option,       <chr>, 1       msg, 2      flow, 3   content, 4     dept...
    ## $ id        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13

``` r
rule_vars(rules)
```

    ## [1] "$EXTERNAL_NET"   "$HOME_NET"       "$TELNET_SERVERS"

``` r
glimpse(rules$options[[1]])
```

    ## Observations: 9
    ## Variables: 2
    ## $ option <chr> "msg", "flow", "content", "depth", "reference", "classtype", "sid", "rev", "metadata"
    ## $ value  <chr> "\"ET TELNET External Telnet Attempt To Cisco Device With No Telnet Password Set (Automatically Diss...

``` r
rules$options[[1]]
```

    ## # A tibble: 9 x 2
    ##      option
    ##       <chr>
    ## 1       msg
    ## 2      flow
    ## 3   content
    ## 4     depth
    ## 5 reference
    ## 6 classtype
    ## 7       sid
    ## 8       rev
    ## 9  metadata
    ## # ... with 1 more variables: value <chr>

## A slightly bigger example

Let’s slurp in all the [Emerging
Threats](https://rules.emergingthreats.net/) Snort feed rules.

Grabbed & unpacked from:
<https://rules.emergingthreats.net/open/snort-edge/emerging.rules.tar.gz>.

``` r
list.files("rules/emerging-rules", "\\.rules$", full.names=TRUE) %>% 
  map_df(~{
    cat(crayon::green(.x), "\n", sep="")
    x <- read_rules(.x)
    if (!is.null(x)) mutate(x, fil = .x)
  }) %>% as_rule_df()-> xdf
```

    ## rules/emerging-rules/emerging-activex.rules
    ## rules/emerging-rules/emerging-attack_response.rules
    ## rules/emerging-rules/emerging-botcc.portgrouped.rules
    ## rules/emerging-rules/emerging-botcc.rules
    ## rules/emerging-rules/emerging-chat.rules
    ## rules/emerging-rules/emerging-ciarmy.rules
    ## rules/emerging-rules/emerging-compromised.rules
    ## rules/emerging-rules/emerging-current_events.rules
    ## rules/emerging-rules/emerging-deleted.rules
    ## rules/emerging-rules/emerging-dns.rules
    ## rules/emerging-rules/emerging-dos.rules
    ## rules/emerging-rules/emerging-drop.rules
    ## rules/emerging-rules/emerging-dshield.rules
    ## rules/emerging-rules/emerging-exploit.rules
    ## rules/emerging-rules/emerging-ftp.rules
    ## rules/emerging-rules/emerging-games.rules
    ## rules/emerging-rules/emerging-icmp_info.rules
    ## rules/emerging-rules/emerging-icmp.rules
    ## rules/emerging-rules/emerging-imap.rules
    ## rules/emerging-rules/emerging-inappropriate.rules
    ## rules/emerging-rules/emerging-info.rules
    ## rules/emerging-rules/emerging-malware.rules
    ## rules/emerging-rules/emerging-misc.rules
    ## rules/emerging-rules/emerging-mobile_malware.rules
    ## rules/emerging-rules/emerging-netbios.rules
    ## rules/emerging-rules/emerging-p2p.rules
    ## rules/emerging-rules/emerging-policy.rules
    ## rules/emerging-rules/emerging-pop3.rules
    ## rules/emerging-rules/emerging-rbn-malvertisers.rules
    ## rules/emerging-rules/emerging-rbn.rules
    ## rules/emerging-rules/emerging-rpc.rules
    ## rules/emerging-rules/emerging-scada.rules
    ## rules/emerging-rules/emerging-scan.rules
    ## rules/emerging-rules/emerging-shellcode.rules
    ## rules/emerging-rules/emerging-smtp.rules
    ## rules/emerging-rules/emerging-snmp.rules
    ## rules/emerging-rules/emerging-sql.rules
    ## rules/emerging-rules/emerging-telnet.rules
    ## rules/emerging-rules/emerging-tftp.rules
    ## rules/emerging-rules/emerging-tor.rules
    ## rules/emerging-rules/emerging-trojan.rules
    ## rules/emerging-rules/emerging-user_agents.rules
    ## rules/emerging-rules/emerging-voip.rules
    ## rules/emerging-rules/emerging-web_client.rules
    ## rules/emerging-rules/emerging-web_server.rules
    ## rules/emerging-rules/emerging-web_specific_apps.rules
    ## rules/emerging-rules/emerging-worm.rules

``` r
glimpse(xdf)
```

    ## Observations: 26,155
    ## Variables: 11
    ## $ action    <chr> "alert", "alert", "alert", "alert", "alert", "alert", "alert", "alert", "alert", "alert", "alert"...
    ## $ protocol  <chr> "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp", "tcp",...
    ## $ src_addr  <chr> "$EXTERNAL_NET", "$EXTERNAL_NET", "$EXTERNAL_NET", "$EXTERNAL_NET", "$EXTERNAL_NET", "$EXTERNAL_N...
    ## $ src_ports <chr> "$HTTP_PORTS", "$HTTP_PORTS", "$HTTP_PORTS", "$HTTP_PORTS", "$HTTP_PORTS", "$HTTP_PORTS", "$HTTP_...
    ## $ direction <chr> "unidirectional", "unidirectional", "unidirectional", "unidirectional", "unidirectional", "unidir...
    ## $ dst_addr  <chr> "$HOME_NET", "$HOME_NET", "$HOME_NET", "$HOME_NET", "$HOME_NET", "$HOME_NET", "$HOME_NET", "$HOME...
    ## $ dst_ports <chr> "any", "any", "any", "any", "any", "any", "any", "any", "any", "any", "any", "any", "any", "any",...
    ## $ commented <lgl> FALSE, TRUE, TRUE, FALSE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, FALSE, FALS...
    ## $ options   <list> [<# A tibble: 20 x 2,       option,        <chr>,  1       msg,  2      flow,  3   content,  4  ...
    ## $ id        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27...
    ## $ fil       <chr> "rules/emerging-rules/emerging-activex.rules", "rules/emerging-rules/emerging-activex.rules", "ru...

``` r
xdf
```

    ## # A tibble: 26,155 x 11
    ##    action protocol      src_addr   src_ports      direction  dst_addr dst_ports commented           options    id
    ##     <chr>    <chr>         <chr>       <chr>          <chr>     <chr>     <chr>     <lgl>            <list> <int>
    ##  1  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any     FALSE <tibble [20 x 2]>     1
    ##  2  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any      TRUE <tibble [23 x 2]>     2
    ##  3  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any      TRUE <tibble [16 x 2]>     3
    ##  4  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any     FALSE <tibble [28 x 2]>     4
    ##  5  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any      TRUE <tibble [16 x 2]>     5
    ##  6  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any      TRUE <tibble [16 x 2]>     6
    ##  7  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any      TRUE <tibble [16 x 2]>     7
    ##  8  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any      TRUE <tibble [16 x 2]>     8
    ##  9  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any      TRUE <tibble [16 x 2]>     9
    ## 10  alert      tcp $EXTERNAL_NET $HTTP_PORTS unidirectional $HOME_NET       any      TRUE <tibble [16 x 2]>    10
    ## # ... with 26,145 more rows, and 1 more variables: fil <chr>

What are the most referenced URLs in Emerging Threats feed?

``` r
unnest(xdf) %>% 
  filter(option == "reference") %>% 
  filter(grepl("^url", value)) %>% 
  select(value) %>% 
  mutate(value = sub("^url,", "", value)) %>% 
  count(value, sort=TRUE)
```

    ## # A tibble: 14,425 x 2
    ##                                                                          value     n
    ##                                                                          <chr> <int>
    ##  1                              doc.emergingthreats.net/bin/view/Main/TorRules  1452
    ##  2                                 doc.emergingthreats.net/bin/view/Main/BotCC   657
    ##  3                                                              sslbl.abuse.ch   608
    ##  4 blog.unmaskparasites.com/2012/06/22/runforestrun-and-pseudo-random-domains/   407
    ##  5        blog.opendns.com/2012/07/10/opendns-security-team-blackhole-exploit/   401
    ##  6                                                  ransomwaretracker.abuse.ch   358
    ##  7                                                        www.shadowserver.org   261
    ##  8                 www.appsecinc.com/resources/alerts/oracle/2004-0001/25.html   211
    ##  9                                                           www.cinsscore.com   200
    ## 10                                                www.networkcloaking.com/cins   200
    ## # ... with 14,415 more rows

## Extended v2 Reading

``` r
evt <- read_extended(system.file("extdata", "multi-record-event-x2.log", package="porc"))

dplyr::glimpse(evt)
```

    ## Observations: 34
    ## Variables: 32
    ## $ sensor_id           <int> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
    ## $ event_id            <int> 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89, 89,...
    ## $ event_second        <int> 964798804, 964798804, 964798804, 964798804, 964798804, 964798804, 964798804, 964798804,...
    ## $ event_microsecond   <int> 267362, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 267362, NA, NA,...
    ## $ signature_id        <int> 3, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 3, NA, NA, NA, NA, N...
    ## $ generator_id        <int> 120, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 120, NA, NA, NA, N...
    ## $ signature_revision  <int> 1, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 1, NA, NA, NA, NA, N...
    ## $ classification_id   <int> 2, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 2, NA, NA, NA, NA, N...
    ## $ priority            <int> 3, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 3, NA, NA, NA, NA, N...
    ## $ source_ip_raw       <list> [<cf, 19, 47, 1c>, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, N...
    ## $ destination_ip_raw  <list> [<0a, 14, 0b, 7b>, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, N...
    ## $ sport_itype         <int> 80, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 80, NA, NA, NA, NA,...
    ## $ dport_icode         <int> 2651, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 2651, NA, NA, NA,...
    ## $ protocol            <int> 6, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 6, NA, NA, NA, NA, N...
    ## $ impact_flag         <int> 0, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 0, NA, NA, NA, NA, N...
    ## $ impact              <int> 0, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 0, NA, NA, NA, NA, N...
    ## $ blocked             <int> 0, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 0, NA, NA, NA, NA, N...
    ## $ mpls_label          <int> 0, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 0, NA, NA, NA, NA, N...
    ## $ vlan_id             <int> 0, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 0, NA, NA, NA, NA, N...
    ## $ pad2                <int> 0, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 0, NA, NA, NA, NA, N...
    ## $ rec_type            <chr> "event_v2", "extra_data", "packet", "packet", "packet", "packet", "packet", "packet", "...
    ## $ rec_num             <dbl> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, ...
    ## $ event_type          <int> NA, 4, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 4, NA, NA, NA, N...
    ## $ event_length        <int> NA, 18602, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 18602, NA, N...
    ## $ type                <int> NA, 13, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 13, NA, NA, NA,...
    ## $ data_type           <int> NA, 1, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 1, NA, NA, NA, N...
    ## $ data_length         <int> NA, 18578, 227, 1163, 820, 1084, 1514, 1514, 1514, 1514, 1514, 1298, 1514, 1514, 1514, ...
    ## $ data                <list> [NULL, <3c, 48, 54, 4d, 4c, 3e, 0a, 3c, 48, 45, 41, 44, 3e, 0a, 09, 3c, 54, 49, 54, 4c...
    ## $ extra_data_type_map <chr> NA, "NORMALIZED_JS", NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, "N...
    ## $ packet_second       <int> NA, NA, 964798804, 964798804, 964798804, 964798804, 964798804, 964798804, 964798804, 96...
    ## $ packet_microsecond  <int> NA, NA, 267362, 276118, 279940, 365865, 374238, 382123, 390039, 482792, 490652, 497184,...
    ## $ link_type           <int> NA, NA, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, NA, NA, 1, 1, 1, 1, 1, 1, 1, 1, 1,...

## Expanding Snort Port Ranges

``` r
rng_ex <- c("25", "$HTTP_PORTS", "1024:", ":1024", "1:1024", "any")
rng <- expand_port_ranges(rng_ex)

str(
  setNames(
    list(
      rng[[1]],
      rng[[2]],
      range(as.numeric(rng[[3]])),
      range(as.numeric(rng[[4]])),
      range(as.numeric(rng[[5]])),
      rng[[6]]
    ),
    rng_ex
  )
)
```

    ## List of 6
    ##  $ 25         : chr "25"
    ##  $ $HTTP_PORTS: chr "$HTTP_PORTS"
    ##  $ 1024:      : num [1:2] 1024 65535
    ##  $ :1024      : num [1:2] 1 1024
    ##  $ 1:1024     : num [1:2] 1 1024
    ##  $ any        : chr "any"

## Code of Conduct

Please note that this project is released with a [Contributor Code of
Conduct](CONDUCT.md). By participating in this project you agree to
abide by its terms.


NEWS.md
0.1.1

* Extended v2 log reading partial support (WIP)

0.1.0 

* Initial release


tests/test-all.R
library(testthat)
test_check("porc")


tests/testthat/test-porc.R
context("read_rules() functions properly")
test_that("we can read rules", {

  rules <- read_rules(system.file("extdata", "emerging-telnet.rules", package="porc"))

  expect_equal(nrow(rules), 13)
  expect_equal(ncol(rules), 10)
  expect_that(rules, is_a("snort_rule_df"))

})

context("rule_vars() functions properly")
test_that("we can read rules", {

  rules <- read_rules(system.file("extdata", "emerging-telnet.rules", package="porc"))

  expect_equal(rule_vars(rules), c("$EXTERNAL_NET", "$HOME_NET", "$TELNET_SERVERS"))

})

context("rule_vars() functions properly")
test_that("we can read extended v2 logs", {

  evt <- read_extended(system.file("extdata", "multi-record-event-x2.log", package="porc"))

  expect_equal(dim(evt), c(34L, 32L))

})

context("expand_ranges() functions properly")
test_that("we can expand ranges", {

  rnd <- expand_port_ranges(c("25", "$HTTP_PORTS", "1024:", ":1024", "1:1024", "any"))

  expect_equal(rnd[[1]], "25")
  expect_equal(rnd[[2]], "$HTTP_PORTS")
  expect_equal(rnd[[3]][[10]], "1033")
  expect_equal(rnd[[4]][[100]], "100")
  expect_equal(rnd[[5]][30], "30")
  expect_equal(rnd[[6]], "any")

})



R/read-rules.R
#' Parse in a file of snort rules into a data frame
#'
#' The individual components of a Snort rule are parsed and put into a tidy
#' data frame format. The rule options are also parsed and the resultant
#' `options` columns is a data frame with option name and value. The value
#' for options that do not have values is `NA_character`.\cr
#' \cr
#' There is an `id` column which is just an index of the rule position in the
#' file. An extra `commented` field exists to identify rules that are in the
#' file but commented out. This makes it easy to filter on in-use rules.\cr
#' \cr
#' The `options` column can be unnested with `tidyr::unnest()`.
#'
#' @md
#' @param path path to snort rules file
#' @return data frame
#' @export
#' @examples
#' rules <- read_rules(
#'   system.file("extdata", "emerging-telnet.rules", package="porc")
#' )
read_rules <- function(path) {

  path <- normalizePath(path.expand(path))
  if (!file.exists(path)) stop("File not found", call.=FALSE)

  l <- stri_read_lines(path)
  # l <- suppressWarnings(stri_enc_toutf8(l, is_unknown_8bit = TRUE, validate = TRUE))
  l <- stri_trim_both(l)
  l <- Filter(not_blank, l)

  # Everything up to the first left parens is the "header" of the rule
  # The parenthetical section is made up of options that are semicolon-separated

  x <- stri_split_fixed(l, "(", 2, simplify=TRUE)
  x <- as.data.frame(x, stringsAsFactors=FALSE)
  x$V2 <- ifelse(x$V2 == "", NA_character_, x$V2)
  x$V1 <- ifelse(grepl(">", x$V1, fixed=TRUE), x$V1, NA_character_)
  x <- x[complete.cases(x),]
  x <- x[grepl(")", x$V2),]

  if (nrow(x) == 0) return(NULL) # No rules

  x$V2 <- sub(")$", "", x$V2)

  commented <- grepl("^#[[:space:]]*", x$V1)

  x$V1 <- sub("^#[[:space:]]*", "", x$V1)

  as.data.frame(
    stri_split_regex(x$V1, "[[:space:]]+", simplify=TRUE),
    stringsAsFactors=FALSE
  ) -> rule_df

  setNames(rule_df,
    c("action", "protocol", "src_addr", "src_ports", "direction",
      "dst_addr", "dst_ports", "activatedynamic")) -> rule_df

  rule_df <- rule_df[complete.cases(rule_df),]

  rule_df$activatedynamic <- NULL

  rule_df$commented <- commented

  rule_df$direction <- c("->"="unidirectional", "<>"="bidirectional")[rule_df$direction]

  rule_df$options <- lapply(
    stri_split_regex(x$V2, ";[[:space:]]*"),
    function(.x) {
      Filter(not_blank, stri_trim_both(.x))
    }
  )

  rule_df$options <- lapply(rule_df$options, function(.x) {

    noc <- Filter(no_colon, .x)

    .x <- Filter(has_colon, .x)
    .x <- stri_split_fixed(.x, ":", simplify=TRUE)
    .x <- as.data.frame(.x, stringsAsFactors=FALSE)

    class(.x) <- c("tbl_df", "tbl", "data.frame")

    if (length(.x) > 0) {
      .x <- setNames(.x, c("option", "value"))
      .x <- .x[,1:2]
    }

    if (length(noc) > 0) {
      .y <- data.frame(option = noc, value=NA_character_, stringsAsFactors=FALSE)
      if (length(.x) > 0) {
        .x <- rbind(.x, .y)
      } else {
        .x <- .y
      }
    }

    .x

  })

  rule_df$id <- 1:nrow(rule_df)
  rule_df <- as_rule_df(rule_df)

  as_rule_df(rule_df[,1:10])

}



R/expand-range.R
sprintf("(%s)", paste0(c(
  "^(?<var>\\$[[:alpha:][:digit:]_]+)$",
  "(?<any>any)",
  "^(?<single>[[:digit:]]+)$",
  "(?<l>[[:digit:]]+)\\:(?<r>[[:digit:]]+)",
  "^\\:(?<to>[[:digit:]]+)",
  "(?<from>[[:digit:]]+)\\:$"
), collapse = "|")) -> snort_ranges_regex_text


#' Expand a Snort port description
#'
#' Snort rule port descriptions can be anything from a single number, to `any`,
#' to a `$` variable specifier to a full or partial range. This function will
#' identify full or partial ranges and expand then, leaving other port descriptors
#' untouched.
#'
#' @md
#' @return list
#' @param x character vector of individual port strings
#' @export
#' @examples
#' expand_port_ranges(c("25", "$HTTP_PORTS", "1024:", ":1024", "1:1024", "any"))
expand_port_ranges <- function(x) {

  snort_ranges_regex <- ore::ore(snort_ranges_regex_text)

  y <- ore::groups(ore::ore.search(snort_ranges_regex, x, simplify=FALSE))
  z <- dplyr::as_data_frame(t(y[,,]))

  purrr::map(1:nrow(z), ~{
    zq <- z[.x,]
    if (!(is.na(zq$single))) { return(zq$single) }
    if (!(is.na(zq$var))) { return(zq$var) }
    if (!(is.na(zq$any))) { return("any") }
    if (complete.cases(zq[,c("l", "r")])) { return(as.character(seq(as.numeric(zq$l), as.numeric(zq$r), 1))) }
    if (!(is.na(zq$to))) { return(as.character(seq(1, as.numeric(zq$to), 1))) }
    if (!(is.na(zq$from))) { return(as.character(seq(as.numeric(zq$from), 65535, 1))) }
    return(NA_character_)
  })

}


R/extendedv2.R
app_id_length <-  16

valid_record_types <- list(
  packet = 2,
  event = 7,
  event_ipv6 = 72,
  event_v2 = 104,
  event_ipv6_v2 = 105,
  extra_data = 110,
  event_appid = 111,
  event_appid_ipv6 = 112,
  appstat = 113
)

record_type_map <- list(
  `2` = "packet",
  `7` = "event",
  `72` = "event_ipv6",
  `104` = "event_v2",
  `105` = "event_ipv6_v2",
  `110` = "extra_data",
  `111` = "event_appid",
  `112` = "event_appid_ipv6",
  `113` = "appstat"
)

extra_data_type <- list(
  `1` = "XFF_IP4",
  `2` = "XFF_IP6",
  `3` = "REVIEWED_BY",
  `4` = "GZIP_DATA",
  `5` = "SMTP_FILENAME",
  `6` = "SMTP_MAIL_FROM",
  `7` = "SMTP_RCPT_TO",
  `8` = "SMTP_HEADERS",
  `9` = "HTTP_URI",
  `10` = "HTTP_HOSTNAME",
  `11` = "IP6_SRC_ADDR",
  `12` = "IP6_DST_ADDR",
  `13` = "NORMALIZED_JS"
)

packet_fields  <- list(
  "sensor_id" = list(what="integer", n=1, size=4),
  "event_id" = list(what="integer", n=1, size=4),
  "event_second" = list(what="integer", n=1, size=4),
  "packet_second" = list(what="integer", n=1, size=4),
  "packet_microsecond" = list(what="integer", n=1, size=4),
  "link_type" = list(what="integer", n=1, size=4),
  "data_length" = list(what="integer", n=1, size=4)
)

event_fields <-  list(
  "sensor_id" = list(what="integer", n=1, size=4),
  "event_id" = list(what="integer", n=1, size=4),
  "event_second" = list(what="integer", n=1, size=4),
  "event_microsecond" = list(what="integer", n=1, size=4),
  "signature_id" = list(what="integer", n=1, size=4),
  "generator_id" = list(what="integer", n=1, size=4),
  "signature_revision" = list(what="integer", n=1, size=4),
  "classification_id" = list(what="integer", n=1, size=4),
  "priority" = list(what="integer", n=1, size=4),
  "source_ip_raw" = list(what = "raw", n=4, size=1),
  "destination_ip_raw" = list(what = "raw", n=4, size=1),
  "sport_itype" = list(what = "integer", n=1, size=2),
  "dport_icode" = list(what = "integer", n=1, size=2),
  "protocol" = list(what = "integer", n=1, size=1),
  "impact_flag" = list(what = "integer", n=1, size=1),
  "impact" = list(what = "integer", n=1, size=1),
  "blocked" = list(what = "integer", n=1, size=1)
)

event_v2_fields <- append(
  event_fields,
  list(
    "mpls_label" = list(what = "integer", n=1, size=4),
    "vlan_id" = list(what = "integer", n=1, size=2),
    "pad2" = list(what = "integer", n=1, size=2)
  )
)

event_ipv6_fields <- list(
  "sensor_id" = list(what = "integer", n=1, size=4),
  "event_id" = list(what = "integer", n=1, size=4),
  "event_second" = list(what = "integer", n=1, size=4),
  "event_microsecond" = list(what = "integer", n=1, size=4),
  "signature_id" = list(what = "integer", n=1, size=4),
  "generator_id" = list(what = "integer", n=1, size=4),
  "signature_revision" = list(what = "integer", n=1, size=4),
  "classification_id" = list(what = "integer", n=1, size=4),
  "priority" = list(what = "integer", n=1, size=4),
  "source_ip_raw" = list(what = "raw", n=16, size=1),
  "destination_ip_raw" = list(what = "raw", n=16, size=1),
  "sport_itype" = list(what = "integer", n=1, size=2),
  "dport_icode" = list(what = "integer", n=1, size=2),
  "protocol" = list(what = "integer", n=1, size=1),
  "impact_flag" = list(what = "integer", n=1, size=1),
  "impact" = list(what = "integer", n=1, size=1),
  "blocked" = list(what = "integer", n=1, size=1)
)

event_v2_ipv6_fields <- append(
  event_ipv6_fields,
  list(
    "mpls_label" = list(what = "integer", n=1, size=4),
    "vlan_id" = list(what = "integer", n=1, size=2),
    "pad2" = list(what = "integer", n=1, size=2)
  )
)

extra_data_fields <- list(
  "event_type" = list(what = "integer", n = 1, size = 4),
  "event_length" = list(what = "integer", n = 1, size = 4),
  "sensor_id" = list(what = "integer", n = 1, size = 4),
  "event_id" = list(what = "integer", n = 1, size = 4),
  "event_second" = list(what = "integer", n = 1, size = 4),
  "type" = list(what = "integer", n = 1, size = 4),
  "data_type" = list(what = "integer", n = 1, size = 4),
  "data_length" = list(what = "integer", n = 1, size = 4)
)

list(
  `2` = packet_fields,
  `7` = event_fields,
  `72` = event_ipv6_fields,
  `104` = event_v2_fields,
  `106` = event_v2_ipv6_fields,
  `110` = extra_data_fields#,
  # `111` = event_appid_fields,
  # `112` = event_appid_ipv6_fields,
  # `113` = appstat_fields
) -> field_map

read_fields <- function(con, type, length) {

  buf <- readBin(con, what = "raw", n = length, size =1)

  rec_con <- rawConnection(buf)

  rec <- field_map[[as.character(type)]]

  purrr::map(rec, ~{
    readBin(rec_con, what = .x$what, n = .x$n, size = .x$size, endian="big")
  }) -> r

  if (type %in% c(valid_record_types$extra_data, valid_record_types$packet)) {
    r$data <- list(readBin(rec_con, what="raw", size=1, n=r$data_length))
  }

  if (type %in% c(valid_record_types$extra_data)) {
    r$extra_data_type_map <- extra_data_type[[r$type]]
  }

  r$rec_type <- record_type_map[[as.character(type)]]

  if ("source_ip_raw" %in% names(r)) r$source_ip_raw <- list(r$source_ip_raw)
  if ("destination_ip_raw" %in% names(r)) r$destination_ip_raw <- list(r$destination_ip_raw)

  # iptools::numeric_to_ip(to_int(f$`source-ip.raw`))
  # iptools::numeric_to_ip(to_int(f$`destination-ip.raw`))

  close(rec_con)

  r

}

read_header <- function(con) {
  buf <- readBin(con, what = integer(), n = 2, endian = "big")
  buf <- if (length(buf) > 0) buf <- as.list(setNames(buf, c("type", "length"))) else NULL
  buf
}

#' Read a Snort "extended v2" log into a data frame
#'
#' @param path path to log file
#' @return data frame
#' @export
#' @examples
#' read_extended(system.file("extdata", "multi-record-event-x2.log", package="porc"))
read_extended <- function(path) {

  path <- normalizePath(path.expand(path))

  sz <- file.size(path)
  guess_rec_count <- round(sz/2000)

  con <- file(path, open = "rb")
  ofs <- seek(con, seek(con))

  log_recs <- vector(mode = "list", length = guess_rec_count)

  idx <- 1

  while(!is.null(hdr <- read_header(con))) {
    f <- read_fields(con, hdr$type, hdr$length)
    f$rec_num <- idx
    log_recs[[idx]] <- f
    idx <- idx + 1
  }

  close(con)

  dplyr::bind_rows(log_recs[1:(idx-1)])

}
